pkg/apis/pipeline/v1beta1/task_validation.go

```diff
@@ -571,31 +571,32 @@ func validateNameFormat(stringAndArrayParams sets.String, objectParams []ParamSp
 }
 
 func validateStepVariables(ctx context.Context, step Step, prefix string, vars sets.String) *apis.FieldError {
-	errs := validateTaskVariable(step.Name, prefix, vars).ViaField("name")
-	errs = errs.Also(validateTaskVariable(step.Image, prefix, vars).ViaField("image"))
-	errs = errs.Also(validateTaskVariable(step.WorkingDir, prefix, vars).ViaField("workingDir"))
-	if !(config.FromContextOrDefaults(ctx).FeatureFlags.EnableAPIFields == "alpha" && prefix == "params") {
-		errs = errs.Also(validateTaskVariable(step.Script, prefix, vars).ViaField("script"))
-	}
+	errs := validateTaskVariable(ctx, step.Name, prefix, vars).ViaField("name")
+	errs = errs.Also(validateTaskVariable(ctx, step.Image, prefix, vars).ViaField("image"))
+	errs = errs.Also(validateTaskVariable(ctx, step.WorkingDir, prefix, vars).ViaField("workingDir"))
+	errs = errs.Also(validateTaskVariable(ctx, step.Script, prefix, vars).ViaField("script"))
 	for i, cmd := range step.Command {
-		errs = errs.Also(validateTaskVariable(cmd, prefix, vars).ViaFieldIndex("command", i))
+		errs = errs.Also(validateTaskVariable(ctx, cmd, prefix, vars).ViaFieldIndex("command", i))
 	}
 	for i, arg := range step.Args {
-		errs = errs.Also(validateTaskVariable(arg, prefix, vars).ViaFieldIndex("args", i))
+		errs = errs.Also(validateTaskVariable(ctx, arg, prefix, vars).ViaFieldIndex("args", i))
 	}
 	for _, env := range step.Env {
-		errs = errs.Also(validateTaskVariable(env.Value, prefix, vars).ViaFieldKey("env", env.Name))
+		errs = errs.Also(validateTaskVariable(ctx, env.Value, prefix, vars).ViaFieldKey("env", env.Name))
 	}
 	for i, v := range step.VolumeMounts {
-		errs = errs.Also(validateTaskVariable(v.Name, prefix, vars).ViaField("name").ViaFieldIndex("volumeMount", i))
-		errs = errs.Also(validateTaskVariable(v.MountPath, prefix, vars).ViaField("MountPath").ViaFieldIndex("volumeMount", i))
-		errs = errs.Also(validateTaskVariable(v.SubPath, prefix, vars).ViaField("SubPath").ViaFieldIndex("volumeMount", i))
+		errs = errs.Also(validateTaskVariable(ctx, v.Name, prefix, vars).ViaField("name").ViaFieldIndex("volumeMount", i))
+		errs = errs.Also(validateTaskVariable(ctx, v.MountPath, prefix, vars).ViaField("MountPath").ViaFieldIndex("volumeMount", i))
+		errs = errs.Also(validateTaskVariable(ctx, v.SubPath, prefix, vars).ViaField("SubPath").ViaFieldIndex("volumeMount", i))
 	}
 	return errs
 }
 
-func validateTaskVariable(value, prefix string, vars sets.String) *apis.FieldError {
-	return substitution.ValidateVariableP(value, prefix, vars)
+func validateTaskVariable(ctx context.Context, value, prefix string, vars sets.String) *apis.FieldError {
+	if !(config.FromContextOrDefaults(ctx).FeatureFlags.EnableAPIFields == "alpha" && prefix == "params") {
+		return substitution.ValidateVariableP(value, prefix, vars)
+	}
```

Comment:
Yes, so I have a way of validating an inline-spec with propagating params. The solution would be to do the validation in `taskrun_validation` on the `taskrun` instead of `task_validation` on the `taskspec`. For inline spec, all the information is available through the `taskrun` so we can do the validation right there. 

However, for referenced resources where we define `task` separately and pipeline/taskrun has a reference to it, there is no way for the validation to know if the original spec was inline or contained references and so it will skip the validation there.

Reply:
> Yes, so I have a way of validating an inline-spec with propagating params. The solution would be to do the validation in `taskrun_validation` on the `taskrun` instead of `task_validation` on the `taskspec`. For inline spec, all the information is available through the `taskrun` so we can do the validation right there.

Sounds reasonable. So with what you proposed, what will happen in `task_validation`? Will they still validate variable references against params declared in spec? If so, since the spec has no declared params (argument `vars` will be empty), the validation will fail right?

Reply:
yeah the problem seems is that we want to validate params in `task`, with `propagating` feature it can allow us not setting `params` in `task`, they previous validation will fall for this case. And this should not be skipped based on alpha feature flag. So I wonder if having a new flag for this feature makes sense

Reply:
> > Yes, so I have a way of validating an inline-spec with propagating params. The solution would be to do the validation in `taskrun_validation` on the `taskrun` instead of `task_validation` on the `taskspec`. For inline spec, all the information is available through the `taskrun` so we can do the validation right there.
> 
> Sounds reasonable. So with what you proposed, what will happen in `task_validation`? Will they still validate variable references against params declared in spec? If so, since the spec has no declared params (argument `vars` will be empty), the validation will fail right?

Yes, exactly. That's why we cannot validate in `task`.

Reply:
@chuangw6 @Yongxuanzhang here is an update. 

I figured out a way, we can keep all the validations in place with/without propagating parameters. 

When a `taskrun` is provided, it tries to validate the steps (in case they are implicitly defined) via the `taskrunspec`. In this case, it skips validating a `taskSpec` because its already validated via the `taskRunSpec`.

In case a `task` is provided separately and referenced in the taskrun, it will validate the `taskSpec` and not skip over it.

The details are in the implementation.

Reply:
I am wondering if it is a valid case that some params (i.e. `a`, `b`, `c`) are propagated from taskrun and others (i.e. `d`, `e`) are defined explicitly in taskspec.

In this case, with your approach, are we going to just skip validations against `a`, `b`, `c`, but the validation against `d`, `e` is still retained?

Reply:
No, what will happen is that we will do is attempt to validate `a`, `b` and `c` using `validateInlineParameters` defined in `taskrun_validation.go`. It will also validate `d` and `e` using the regular `task_validation` if the task is referenced.